falcon/cluster/spectrum.py

```diff
@@ -245,10 +274,63 @@ def _to_vector(spectra: List[MsmsSpectrumNb], min_mz: float, bin_size: float) \
     i, j = 0, 1
     for spec in spectra:
         n_peaks_spectra = len(spec.mz)
-        data[i:i + n_peaks_spectra] = spec.intensity
+        data[i : i + n_peaks_spectra] = spec.intensity
         mz = [math.floor((mz - min_mz) / bin_size) for mz in spec.mz]
-        indices[i:i + n_peaks_spectra] = mz
+        indices[i : i + n_peaks_spectra] = mz
         indptr[j] = indptr[j - 1] + n_peaks_spectra
         i += n_peaks_spectra
         j += 1
     return data, indices, indptr
+
+
+def df_row_to_spec(row: pd.Series) -> sus.MsmsSpectrum:
+    """
+    Convert a row from a DataFrame to a `MsmsSpectrum`.
+
+    Parameters
+    ----------
+    row : pd.Series
+        A row from a DataFrame containing the spectrum metadata.
+
+    Returns
+    -------
+    MsmsSpectrum
+        The spectrum object.
+    """
+    spectrum = sus.MsmsSpectrum(
+        identifier=row["identifier"],
+        precursor_mz=row["precursor_mz"],
+        precursor_charge=row["precursor_charge"],
+        retention_time=row["retention_time"],
+        mz=row["mz"],
+        intensity=row["intensity"],
+    )
+    spectrum.filename = row["filename"]
+    return spectrum
+
+
+def spec_to_dict(
```

Comment:
**question:** Where is this used? There's also the namedtuple defined on top that provides a very similar interface.

Reply:
When reading spectra from a file, the `MsmsSpectrum` objects are converted to a dictionary before writing them to the lance file. I could use the named tuple with `_asdict()` instead. But now that I'm thinking about it, it might be a bit redundant to create a `MsmsSpectrum` first and then, almost immediately, convert it to a dictionary.

Reply:
If we don't need to do any processing on the spectrum, then we could just pass it straight through to avoid some extraneous object creation. `MsmsSpectrum` does a few quality checks though (e.g. ensure that the peaks are sorted by m/z), so it might still be useful to avoid sneaky issues. But we should definitely try to avoid unnecessary data copies.

Reply:
It would be nice to keep the quality checks in `MsmsSpectrum`.

Reply:
Removed this function. The processed spectra are now returned as a dict before adding them to the lance dataset.